src/library/dao/trackdao.cpp

```diff
@@ -630,49 +632,69 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
             LOG_FAILED_QUERY(*m_pQueryLibraryInsert)
                     << "Failed to INSERT new track into library:"
                     << pTrack->getLocation();
-            return false;
+            return TrackId();
         }
         trackId = TrackId(m_pQueryLibraryInsert->lastInsertId());
+        m_analysisDao.saveTrackAnalyses(pTrack.data());
+        m_cueDao.saveTrackCues(trackId, pTrack.data());
+        trackInserted = true;
+    }
+    if (trackId.isValid()) {
         pTrack->setId(trackId);
-        m_analysisDao.saveTrackAnalyses(pTrack);
-        m_cueDao.saveTrackCues(trackId, pTrack);
+    } else {
+        qWarning() << "TrackDAO::addTracksAddTrack:"
+                << "Failed to add track to database"
+                << pTrack->getLocation();
+    }
+    // When exiting this function the trackId obtained from the
+    // database must match the id of the track, no matter if the
+    // track has been inserted or updated. If the trackId is still
+    // invalid adding the track to the library must have failed
+    // and the id of the track should also still be invalid.
+    DEBUG_ASSERT(pTrack->getId() == trackId);
+    m_tracksAddedSet.insert(trackId);
+    if (trackInserted) {
         pTrack->markClean();
     }
-    m_tracksAddedSet.insert(trackId);
-    return true;
+    return trackId;
 }
 
-TrackId TrackDAO::addTrack(const QFileInfo& fileInfo, bool unremove) {
+TrackId TrackDAO::addTracksAddFile(const QFileInfo& fileInfo, bool unremove) {
+    // TODO(uklotzde): Resolve TrackInfoObject through TrackCache
     TrackPointer pTrack(new TrackInfoObject(fileInfo));
-    SoundSourceProxy(pTrack).loadTrackMetadata();
-    // Add the song to the database.
-    addTrack(pTrack.data(), unremove);
-    return pTrack->getId();
-}
 
-void TrackDAO::addTrack(TrackInfoObject* pTrack, bool unremove) {
-    //qDebug() << "TrackDAO::addTrack" << QThread::currentThread() << m_database.connectionName();
-    //qDebug() << "TrackCollection::addTrack(), inserting into DB";
-    // Why you be giving me NULL pTracks
-    DEBUG_ASSERT_AND_HANDLE(pTrack) {
-        return;
+    // Check that track is a supported extension.
+    if (!SoundSourceProxy::isFileSupported(fileInfo)) {
+        qWarning() << "TrackDAO::addTracksAddFile:"
+                << "Unsupported file type"
+                << pTrack->getLocation();
+        return TrackId();
     }
 
-    // Check that track is a supported extension.
-    if (!SoundSourceProxy::isFileSupported(pTrack->getFileInfo())) {
-        // TODO(XXX) provide some kind of error code on a per-track basis.
-        return;
+    SoundSourceProxy(pTrack).loadTrackMetadata();
```

Comment:
Here we load the track metadata even though we do not need it in any case. If the track is already in library it is not used. 
It is probably not a topic for this PR, but It looks like we can spit the location insert and the track insert in two functions and call the location insert before loadTrackMetadata()